### ISSUE TITLE
Akka-HTTP textRequest matches any Content-Type

### ISSUE THREAD

**Issue by emanb29** [julienrf/endpoints#534]
[akkahttp.server.EndpointsWithCustomErrors#textRequest  ](https://github.com/julienrf/endpoints/blob/master/akka-http/server/src/main/scala/endpoints/akkahttp/server/Endpoints.scala#L117) matches any `Content-Type` header, while [openapi.Requests#textRequest](https://github.com/julienrf/endpoints/blob/master/openapi/openapi/src/main/scala/endpoints/openapi/Requests.scala#L45) documents the algebra as only matching `text/plain`.

Is this intended behavior? If so, is the rationale documented anywhere?

**Comment by julienrf** [comment]
> Is this intended behavior?

So far, yes, but we can change this :)

> If so, is the rationale documented anywhere?

It is not documented. My motivation for this behavior was to have more tolerant servers: they accept requests that have a wrong or missing `Content-Type` header, but they produce responses with the correct `Content-Type`. Conversely, clients produce requests with the correct `Content-Type`, but don’t pay attention to the responses’ `Content-Type`.

I’m happy to reconsider this decision if it causes more harm than good!

**Comment by harpocrates** [comment]
> It is not documented. My motivation for this behavior was to have more tolerant servers: they accept requests that have a wrong or missing Content-Type header, but they produce responses with the correct Content-Type. Conversely, clients produce requests with the correct Content-Type, but don’t pay attention to the responses’ Content-Type.
>
> I’m happy to reconsider this decision if it causes more harm than good!

This is problematic for the `orElse` proposed in #529 (since the whole point there is to use the content-type to decide on unmarshalling). `textRequest.orElse(jsonRequest[Foo])` ends up always incorrectly processing JSON input since the `textRequest` variant ignores the `application/json` header and always matches anyways.

Perhaps `textRequest` could pay attention to `Content-Type`, but there could be another  weaker variant `uncheckedTextRequest` (with an informative Scaladoc)?

**Comment by julienrf** [comment]
You’re right, implementing `orElse` on request entities requires us to decide which branch to pick according to the `Content-Type`. So, indeed, we should change that.

What do you think of the following? If the `Content-Type` header is present we validate that it matches the one defined by the `RequestEntity`, but if the header is missing, it’s fine (we still try to decode the request entity).

As you suggested, we could have something like `anyRequest: RequestEntity[String]` which would accept any content type. It could be a `RequestEntity[Array[Byte]]`, actually.

**Comment by harpocrates** [comment]
I've done some more reading since my last comment and my position has shifted a bit to appreciate more the forgiving `textRequest`.

> As you suggested, we could have something like anyRequest: RequestEntity[String] which would accept any content type. It could be a RequestEntity[Array[Byte]], actually.

The current `textRequest` behaviour is more complex than that, and probably much more useful for "forgiving servers". At least in the Akka HTTP server interpreter, it will match _any requests that has a content-type that can be decoded as a string_ (ex: set through the [`charset` directive](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Type#Directives)), doing [some useful work to pick the right encoding](https://github.com/akka/akka-http/blob/6de639f641880576e089518d8ba000c32e9f8e62/akka-http/src/main/scala/akka/http/scaladsl/unmarshalling/PredefinedFromEntityUnmarshallers.scala#L33-L37).

So how about:

  * `plainTextRequest` is added and accepts only content-type `text/plain`
  * `textRequest` maintains its current behaviour of accepting any request that can be decoded as text, but we extend its Scaladoc to talk about its limitations (doesn't play well with `orElse`, leads to imprecise OpenAPI documentation) and recommend `plainTextRequest` instead when possible

**Comment by emanb29** [comment]
Here's a quick-fix implementation of `plainTextRequest` for the Akka interpreter
```scala
  def plainTextRequest: RequestEntity[String] = {
    implicit val um: FromEntityUnmarshaller[String] =
      Unmarshaller.stringUnmarshaller.forContentTypes(MediaTypes.`text/plain`)

    Directives.entity[String](implicitly)
  }
```

**Comment by julienrf** [comment]
I think this has been fixed by #544?

**Comment by emanb29** [comment]
I think so too 👍
